connectors/slack/template/configure.js

```diff
@@ -47,38 +48,58 @@ module.exports = {
                 const selfUrl = Sdk.getSelfUrl(ctx);
                 let response;
                 try {
-                    response = await Superagent.post(`https://slack.com/api/oauth.v2.access`)
-                        .type('form')
-                        .send({
-                            code: ctx.query.code,
-                            client_id: ctx.configuration.slack_client_id, 
-                            client_secret: ctx.configuration.slack_client_secret,
-                            redirect_uri: selfUrl
-                        });
+                    response = await Superagent.post(`https://slack.com/api/oauth.v2.access`).type('form').send({
+                        code: ctx.query.code,
+                        client_id: ctx.configuration.slack_client_id,
+                        client_secret: ctx.configuration.slack_client_secret,
+                        redirect_uri: selfUrl,
+                    });
                     if (!response || !response.body || !response.body.access_token) {
                         throw new Error('Unable to obtain access token');
                     }
-                }
-                catch (e) {
-                    throw { 
-                        status: 500, 
-                        message: `Slack authorization failed: ${e.message}`, 
-                        state 
+                } catch (e) {
+                    throw {
+                        status: 500,
+                        message: `Slack authorization failed: ${e.message}`,
+                        state,
                     };
                 }
-                let data = { 
-                    ...state.data, 
-                    slack_access_token: response.body.access_token
+
+                let data = {
+                    ...state.data,
+                    slack_bot_access_token: response.body.access_token,
+                    slack_bot_id: response.body.bot_user_id,
+                    slack_user_access_token: response.body.authed_user && response.body.authed_user.access_token,
+                    slack_user_id: response.body.authed_user && response.body.authed_user.id,
+                    slack_app_id: response.body.app_id,
+                    slack_team_id: response.body.team && response.body.team.id,
                 };
+
+                if (ctx.configuration.slack_signing_secret) {
+                    // Generate and store token the function will later user to call back for
+                    // events registration
+                    const tokenLength = 48;
+                    const tokenUses = 2;
+
+                    storage.ensureStorage(ctx);
+                    await storage.ensureCache();
+                    let token = (await crypto.randomBytes(tokenLength)).toString('hex');
+                    await storage.put(() => {
+                        // Store which app this token is for and 2 remaining uses
```

Comment:
Why 2?  Add some comments here detailing the approach.

Reply:
Fixed